### ISSUE TITLE
Ability for server to connect back to Cassandra when all host moves

### ISSUE THREAD

**Issue by samarabbas** [temporalio/temporal#317]
Gocql client only resolve hosts once on session creation.  This is a problem if you are running Cassandra on kubernetes using persisted volumes as all your Cassandra hosts might move at the same time.  This will result in unavailability of Temporal cluster as it won't be able to connect to Cassandra anymore.

Current view of Kubernetes cluster:
```
NAME                                          READY   STATUS    RESTARTS   AGE
pod/cadence-cassandra-0                       1/1     Running   0          25m
pod/cadence-frontend-7bc4f86d8c-vtpvx         1/1     Running   4          3d22h
pod/cadence-frontend-tunnel-6dfbb958b-tz5pj   1/1     Running   0          3d20h
pod/cadence-history-8459c46456-rs2gr          1/1     Running   4          3d22h
pod/cadence-matching-6c74bd46bd-csbjn         1/1     Running   4          3d22h
pod/cadence-web-555f9c9dbc-ctncs              1/1     Running   0          3d21h
pod/cadence-worker-fffdb84dc-982j8            1/1     Running   0          25m
NAME                                TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                                        AGE
service/cadence-cassandra           ClusterIP   None             <none>        7000/TCP,7001/TCP,7199/TCP,9042/TCP,9160/TCP   3d22h
service/cadence-frontend            ClusterIP   100.67.186.49    <none>        7933/TCP                                       3d22h
service/cadence-frontend-headless   ClusterIP   None             <none>        7933/TCP,9090/TCP                              3d22h
service/cadence-frontend-tunnel     ClusterIP   100.71.154.160   <none>        7933/TCP                                       3d20h
service/cadence-history-headless    ClusterIP   None             <none>        7934/TCP,9090/TCP                              3d22h
service/cadence-matching-headless   ClusterIP   None             <none>        7935/TCP,9090/TCP                              3d22h
service/cadence-web                 ClusterIP   100.70.110.201   <none>        80/TCP                                         3d22h
service/cadence-worker-headless     ClusterIP   None             <none>        7939/TCP,9090/TCP                              3d22h
NAME                                      READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/cadence-frontend          1/1     1            1           3d22h
deployment.apps/cadence-frontend-tunnel   1/1     1            1           3d20h
deployment.apps/cadence-history           1/1     1            1           3d22h
deployment.apps/cadence-matching          1/1     1            1           3d22h
deployment.apps/cadence-web               1/1     1            1           3d22h
deployment.apps/cadence-worker            1/1     1            1           3d22h
NAME                                                DESIRED   CURRENT   READY   AGE
replicaset.apps/cadence-frontend-7bc4f86d8c         1         1         1       3d22h
replicaset.apps/cadence-frontend-tunnel-6dfbb958b   1         1         1       3d20h
replicaset.apps/cadence-history-8459c46456          1         1         1       3d22h
replicaset.apps/cadence-matching-6c74bd46bd         1         1         1       3d22h
replicaset.apps/cadence-web-555f9c9dbc              1         1         1       3d21h
replicaset.apps/cadence-web-57bb9fc7cf              0         0         0       3d22h
replicaset.apps/cadence-web-69795497b8              0         0         0       3d21h
replicaset.apps/cadence-worker-fffdb84dc            1         1         1       3d22h
NAME                                 READY   AGE
statefulset.apps/cadence-cassandra   1/1     3d22h
```

Frontend Logs:
```
{"level":"error","ts":"2020-04-21T16:26:58.340Z","msg":"Operation failed with internal error.","service":"cadence-frontend","error":"gocql: no hosts available in the pool","metric-scope":40,"logging-call-at":"persistenceMetricClients.go:779","stacktrace":"github.com/uber/cadence/common/log/loggerimpl.(*loggerImpl).Error\n\t/cadence/common/log/loggerimpl/logger.go:134\ngithub.com/uber/cadence/common/persistence.(*metadataPersistenceClient).updateErrorMetric\n\t/cadence/common/persistence/persistenceMetricClients.go:779\ngithub.com/uber/cadence/common/persistence.(*metadataPersistenceClient).GetMetadata\n\t/cadence/common/persistence/persistenceMetricClients.go:757\ngithub.com/uber/cadence/common/cache.(*domainCache).refreshDomainsLocked\n\t/cadence/common/cache/domainCache.go:417\ngithub.com/uber/cadence/common/cache.(*domainCache).refreshDomains\n\t/cadence/common/cache/domainCache.go:409\ngithub.com/uber/cadence/common/cache.(*domainCache).refreshLoop\n\t/cadence/common/cache/domainCache.go:393"}
```

Cassandra cluster seems to come up correctly after restart:
```
INFO  [main] 2020-04-21 16:21:02,808 StorageService.java:1446 - JOINING: Finish joining ring
INFO  [main] 2020-04-21 16:21:02,886 SecondaryIndexManager.java:509 - Executing pre-join tasks for: CFS(Keyspace='cadence_visibility', ColumnFamily='schema_update_history')
INFO  [main] 2020-04-21 16:21:02,887 SecondaryIndexManager.java:509 - Executing pre-join tasks for: CFS(Keyspace='cadence_visibility', ColumnFamily='schema_version')
INFO  [main] 2020-04-21 16:21:02,887 SecondaryIndexManager.java:509 - Executing pre-join tasks for: CFS(Keyspace='cadence_visibility', ColumnFamily='closed_executions')
INFO  [main] 2020-04-21 16:21:02,887 SecondaryIndexManager.java:509 - Executing pre-join tasks for: CFS(Keyspace='cadence_visibility', ColumnFamily='open_executions')
INFO  [main] 2020-04-21 16:21:02,888 SecondaryIndexManager.java:509 - Executing pre-join tasks for: CFS(Keyspace='cadence_visibility', ColumnFamily='closed_executions_v2')
INFO  [main] 2020-04-21 16:21:02,888 SecondaryIndexManager.java:509 - Executing pre-join tasks for: CFS(Keyspace='cadence', ColumnFamily='domains')
INFO  [main] 2020-04-21 16:21:02,888 SecondaryIndexManager.java:509 - Executing pre-join tasks for: CFS(Keyspace='cadence', ColumnFamily='history_node')
INFO  [main] 2020-04-21 16:21:02,888 SecondaryIndexManager.java:509 - Executing pre-join tasks for: CFS(Keyspace='cadence', ColumnFamily='queue_metadata')
INFO  [main] 2020-04-21 16:21:02,888 SecondaryIndexManager.java:509 - Executing pre-join tasks for: CFS(Keyspace='cadence', ColumnFamily='events')
INFO  [main] 2020-04-21 16:21:02,889 SecondaryIndexManager.java:509 - Executing pre-join tasks for: CFS(Keyspace='cadence', ColumnFamily='executions')
INFO  [main] 2020-04-21 16:21:02,889 SecondaryIndexManager.java:509 - Executing pre-join tasks for: CFS(Keyspace='cadence', ColumnFamily='history_tree')
INFO  [main] 2020-04-21 16:21:02,889 SecondaryIndexManager.java:509 - Executing pre-join tasks for: CFS(Keyspace='cadence', ColumnFamily='schema_version')
INFO  [main] 2020-04-21 16:21:02,889 SecondaryIndexManager.java:509 - Executing pre-join tasks for: CFS(Keyspace='cadence', ColumnFamily='domains_by_name_v2')
INFO  [main] 2020-04-21 16:21:02,889 SecondaryIndexManager.java:509 - Executing pre-join tasks for: CFS(Keyspace='cadence', ColumnFamily='schema_update_history')
INFO  [main] 2020-04-21 16:21:02,890 SecondaryIndexManager.java:509 - Executing pre-join tasks for: CFS(Keyspace='cadence', ColumnFamily='queue')
INFO  [main] 2020-04-21 16:21:02,890 SecondaryIndexManager.java:509 - Executing pre-join tasks for: CFS(Keyspace='cadence', ColumnFamily='tasks')
INFO  [main] 2020-04-21 16:21:02,973 StorageService.java:2289 - Node /100.104.128.29 state jump to NORMAL
INFO  [main] 2020-04-21 16:21:02,981 Gossiper.java:1692 - Waiting for gossip to settle...
```

**Comment by samarabbas** [comment]
Gocql keeps on trying to connect to old IP of cassandra cluster:
https://github.com/gocql/gocql/blob/68212b7a2cd01dfdd006839a57f77b320a42fc4d/cluster.go#L38

**Comment by samarabbas** [comment]
link to gocql issue:
https://github.com/gocql/gocql/issues/831

**Comment by wxing1292** [comment]
solved by #1342 #1343 #1344 #1345
